internal/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,133 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ratelimit
+
+import (
+	"sync"
+	"time"
+
+	"go.uber.org/yarpc/internal/clock"
+)
+
+// Note: This file is inspired by:
+// https://github.com/prashantv/go-bench/blob/master/ratelimit
+
+// Clock is the minimum necessary interface to instantiate a throttle with a
+// clock or fake clock, compatible with clocks created using
+// github.com/andres-erbsen/clock or go.uber.org/yarpc/internal/clock.
+type Clock interface {
+	Now() time.Time
+}
+
+// Throttle is a rate limiter.
+type Throttle struct {
+	sync.Mutex
+
+	last       time.Time
+	sleepFor   time.Duration
+	perRequest time.Duration
+	maxSlack   time.Duration
+	clock      Clock
+}
+
+// Option is an option for a rate limiter constructor.
+type Option func(l *Throttle)
+
+// NewThrottle returns a Throttle that will limit to the given RPS.
+func NewThrottle(rate int, opts ...Option) *Throttle {
+	l := &Throttle{
+		perRequest: time.Second / time.Duration(rate),
+		maxSlack:   -10 * time.Second / time.Duration(rate),
+	}
+	for _, opt := range opts {
+		opt(l)
+	}
+	if l.clock == nil {
+		l.clock = clock.NewReal()
+	}
+	return l
+}
+
+// WithClock returns an option for ratelimit.New that provides an alternate
+// Clock implementation, typically a mock Clock for testing.
+func WithClock(clock Clock) func(l *Throttle) {
+	return func(l *Throttle) {
+		l.clock = clock
+	}
+}
+
+// WithoutSlack is an option for ratelimit.New that initializes the limiter
+// without any initial tolerance for bursts of traffic.
+func WithoutSlack(l *Throttle) {
+	l.maxSlack = 0
+}
+
+// Throttle returns whether a call should be dropped to ensure that accepted
+// requests remain time.Second/rate on average.
+// All other calls count toward the rate limit.
+func (t *Throttle) Throttle() bool {
+	t.Lock()
+	defer t.Unlock()
+
+	now := t.clock.Now()
+
+	// If this is our first request, then we allow it.
+	if t.last.IsZero() {
+		t.last = now
+		t.sleepFor = t.maxSlack
+		return false
+	}
+
+	// sleepFor calculates how much time we should sleep based on
+	// the perRequest budget and how long the last request took.
+	// Since the request may take longer than the budget, this number
+	// can get negative, and is summed across requests.
+	sleepFor := t.sleepFor + t.perRequest - now.Sub(t.last)
+
+	// If sleepFor is positive, then we should sleep now.
+	if sleepFor >= 0 {
+		return true
+	}
+
+	// We shouldn't allow sleepFor to get too negative, since it would mean that
+	// a service that slowed down a lot for a short period of time would get
+	// a much higher RPS following that.
+	if sleepFor < t.maxSlack {
+		sleepFor = t.maxSlack
```

Comment:
in the initial implementation you referenced,  we used sleepFor to sleep, but we're not doing that in this implementation.  Rename?